[elaboration] Bidirectionality hints

maximedenes opened the pull request:
This feature makes it possible to tell type inference to type
applications of a global `foo` using typing information from the context
once the `n` first arguments are known.

The syntax is: `Arguments foo x y | z`.

Closes #7910.
Closes #5767.

maximedenes commented:
@Zimmi48 I'd like to write some documentation for this, but I failed to find where to start. I could hook this into the documentation of the elaboration and of the general `Arguments` syntax. Can you give a pointer to any of these? Thanks!

maximedenes commented:
I was told by @gares that @ggonthier could be interested too. In particular, I'd welcome any suggestion for a better syntax ;)

Zimmi48 commented:
@maximedenes Most of the current documentation of the elaboration is in the Gallina extensions chapter. In particular, it contains a long section about implicit arguments, canonical structures, and the `Arguments` command. It also contains sections on extensions of `match`, and on Coercions. Thus, it feels like the right place for this, although you'll probably need to start a new section in this chapter because, contrarily to canonical structure, these typing hints do not feel like they belong to the section on implicit arguments. If you want, we can also open an issue about consolidating the documentation on type inference / elaboration.

maximedenes commented:
Thanks for the hints.

> contrarily to canonical structure, these typing hints do not feel like they belong to the section on implicit arguments

Sorry for this naive question, but why would canonical structures belong to a section on implicit arguments?

> If you want, we can also open an issue about consolidating the documentation on type inference / elaboration.

We probably should indeed, because the information is currently hard to find even for developers who kind of know what they are looking for. So I imagine it must be hard for users to discover it.

Zimmi48 commented:
> Sorry for this naive question, but why would canonical structures belong to a section on implicit arguments?

I think it would be better if canonical structures belonged to a section on type inference / elaboration instead. But since it doesn't exist yet, I suppose that's where they fitted best (I don't know much about canonical structures, but if they are liked typeclasses, they serve to resolve implicit arguments automatically, right?).

maximedenes commented:
> they serve to resolve implicit arguments automatically, right?

AFAIK, they serve to infer missing subterms (that are arguments of a canonical projection). Of course, it could come from an implicit argument, or a user-provided `_`. I don't think there's anything specific to implicit arguments there. So it seems strange to put canonical structures in a section about those.

ggonthier commented:
> > they serve to resolve implicit arguments automatically, right?
> AFAIK, they serve to infer missing subterms (that are arguments of a canonical projection).

That's actually not quite correct - CS are hints telling the unification algorithm to handle problems involving a structure projection in a certain way, so they can help infer any missing subterm, not just those that are arguments of the projection. It's fairly common to infer missing proper subterms of arguments of projections, and they sometimes help infer missing sub terms in the term the projection is unified to - possibly even in cases where there are _no_ missing sub terms in the projection call (though the unification heuristics make this somewhat less likely).

maximedenes commented:
> That's actually not quite correct - CS are hints telling the unification algorithm to handle problems involving a structure projection in a certain way, so they can help infer any missing subterm, not just those that are arguments of the projection. It's fairly common to infer missing proper subterms of arguments of projections, and they sometimes help infer missing sub terms in the term the projection is unified to - possibly even in cases where there are _no_ missing sub terms in the projection call (though the unification heuristics make this somewhat less likely).

Ah yes, thanks for the clarification! But it seems still unrelated to implicit arguments, right?

Zimmi48 commented:
I see two options:
1. Create the Elaboration / type inference section now in a separate PR and put canonical structures in it (and maybe coercions and extensions of `match` too, and actually maybe implicit arguments as well as a sub-section), then rebase this PR on top to add bidirectionality hints to this new section.
2. Create this new section in this PR with only bidirectionality hints and add more stuff to it later on.

maximedenes commented:
I'll go for a third option: create a new section in the "Gallina extensions" chapter, and let the current mess being addressed separately. This will avoid delaying this PR because of preexisting doc issues.

ggonthier commented:
> I was told by @gares that @ggonthier could be interested too. In particular, I'd welcome any suggestion for a better syntax ;)

I'm definitely interested in this feature, since I've been whispering about it in Enrico's ear for some years now... One change that should accompany this feature would be to apply it to standard library constants where it is sorely needed, such as `exist`, `exist2`, `existT`, and `existT2`, and indeed possibly implement a facility to automatically insert directionality hints (putting the `|` immediately after the last argument on which the return type depends, when other arguments depend on it), à la `Set Implicit Arguments`.
  In fact, it might be worth looking into the thesis that pushed me to ask for this feature: that Coq's generic unification problem delaying heuristic is only used to improve inference of two things: the `return` predicate of `match`es, and the predicate(s) if `exist` et al. Since directionality hints handle the latter more effectively, if some variant thereof could similarly handle the former, it might be possible to get rid of delayed unification constraints altogether... which Unicoq is attempting, I believe.
  Finally, I have no strong opinions about syntax, but I would note that using `&` instead of `|` would be consistent with the SSReflect alternative syntax for non-dependent arguments in datatype declarations (and more generally telescopes).

maximedenes commented:
> possibly implement a facility to automatically insert directionality hints (putting the `|` immediately after the last argument on which the return type depends, when other arguments depend on it), à la `Set Implicit Arguments`.

Ah ah, this was next on my list ;)

ggonthier commented:
> Ah yes, thanks for the clarification! But it seems still unrelated to implicit arguments, right?
Even more so, though there's the obvious relation that implicit arguments are the primary means of introducing missing sub terms, though by no means the only one (`rewrite`, `(e)apply`, `case`, `Notation`, explicit `_`  and even canonical instance insertion can all do so).

gares commented:
> I'm definitely interested in this feature, since I've been whispering about it in Enrico's ear for some years now...

Well, it turns out last week @maximedenes was prompted to solve an elaboration weakness by the "Coq Consortium support service" and that he came to my office... :grin: I'm very happy the code was finally written!

> it might be possible to get rid of delayed unification constraints altogether... which Unicoq is attempting, I believe.

That is also my hope, but I don't know well enough how Coq typechecks a match (I don't exactly see why delaying there helps its current algorithm) so I'm not so sure but I trust your intuition.

[OT] There is another case where "delay" helps, and that is not related to unification per se but rather the order in which the elaborateor processes terms. Eg `Lemma foo x : x == 1 -> stuff x`. Here `stuff` imposes on `x` a type for which an `eqType` instance exists, but if you process terms in the left to right order than you face the projection-value problem to early to infer a canonical solution. I'm sure this pattern occurs in MathComp, but I also believe it is a quite extreme case and I hope that this can also be solve by "rescheduling toplevel elaboration problems", rather that delaying (potentially deep) unification problems.
Any thoughts on that?

gares commented on pretyping/pretyping.ml:
```diff
@@ -687,30 +716,46 @@ let rec pretype ~program_mode ~poly resolve_tc (tycon : type_constraint) (env :
+    (* We now refine any arguments whose typing was delayed for
+       bidirectionality *)
+    let sigma, t = app in
+    let sigma = List.fold_left refine_arg sigma bidiargs in
```
Aren't you refining these `bidiargs` in the wrong order, w.r.t. the order in which there occur in the original term?

gares replied:
No problem, now I see the `List.rev`

ggonthier commented:
> Eg `Lemma foo x : x == 1 -> stuff x.`
> then you face the projection-value problem too early to infer a canonical solution
Indeed, but this is one area where CS resolution outshines TC: assuming no `Implicit Type` for `x`, the type of `x` is initially `?T : Type`, unification of the constraints for `x == 1` first merely sets `?T := Equality.sort ?eqT` (CS resolution is not called since there is no unification conflict), then CS is used to unify this with the type of `1`, which would be `GRing.Zmodule.sort ?mT`, yielding the assignment `?eqT := GRing.Zmodule.eqType ?mT`, which make `?T` convertible to `GRing.Zmodule.sort ?mT`, and CS resolution may be called again whenever `stuff x` imposes a more precise type for `x`. It might not have to, as first order unification can be used if `stuff x` imposes a type of the form `GRing.Zmodule.sort someZmodType`, and if it never does impose any additional constraints, then TC and/or `Implicit tactic` can step in to fill the gap (and, if you're within an SSreflect proof and are using `have` rather than `Lemma`, then SSReflect will happily regeneralize `?mT` to leave you with a class-polymorphic statement.

maximedenes commented on pretyping/pretyping.ml:
```diff
@@ -635,24 +641,38 @@ let rec pretype ~program_mode ~poly resolve_tc (tycon : type_constraint) (env :
+      try
+        let (gr,_) = destRef sigma fj.uj_val in
+        match GlobRef.Map.find_opt gr !bidi_hints with
+        | Some nargs when nargs >= 0 -> nargs
```
This was a leftover from experimental code. Now fixed, thanks.

maximedenes commented on pretyping/pretyping.ml:
```diff
@@ -687,30 +716,46 @@ let rec pretype ~program_mode ~poly resolve_tc (tycon : type_constraint) (env :
       | _ -> sigma, resj
     in
-    inh_conv_coerce_to_tycon ?loc env sigma resj tycon
+    let app = inh_conv_coerce_to_tycon ?loc env sigma resj tycon in
```
Indeed, fixed, thanks.

maximedenes commented on pretyping/pretyping.ml:
```diff
@@ -662,20 +682,29 @@ let rec pretype ~program_mode ~poly resolve_tc (tycon : type_constraint) (env :
+              typing the argument, so we replace it by an existential
+              variable *)
+              let sigma, uj_val = new_evar env sigma ~src:(loc,Evar_kinds.InternalHole) c1 in
+              (sigma, { uj_val; uj_type = c1 }), (uj_val, c) :: bidiargs
```
Right, thanks for the name suggestion.

maximedenes commented:
> I'd like to see a test like this

Test added, thanks!

> Finally, I have no strong opinions about syntax, but I would note that using `&` instead of `|` would be consistent with the SSReflect alternative syntax for non-dependent arguments in datatype declarations (and more generally telescopes).

I've changed the syntax, thanks for the suggestion!

> Maybe there should be a way to clear the bidir hint for a global too (does putting the `|` at the end of the list have the same effect in general?)

Putting it after the last argument in the application has the same effect, but the last argument at the time of declaring the constant is not always the same (there can be extra arguments at the call site). So I added a `clear bidirectionality hint` modifier for `Arguments`.

> There should be a libobject declaration to allow bidir hints to survive modules/files ;)
> I guess it should behave like other Arguments objects if it's specified through Arguments. I'm not sure if all the Arguments functionality is the same as impargs (require-time, section-local unless `Global` specified).

Of course, where was my mind? I'll try to refactor the libobjects for `Arguments` first, though, as part of a separate PR.

> Also `About`/`Print` should print the bidir hint if present.

Right, I need to do that too.

Janno commented:
Would it make any sense to allow multiple `|`? I imagine the semantics would be that the first `n₁` are used as context for the next `n₂` arguments, then `n₁ + n₂` arguments are used as context for the next group and so on an so forth. It seems like an obvious extension and, of course, exactly because of this obviousness it is at risk of being useless in practice. :)

SkySkimmer commented:
I think the arguments can be discarded in a way we might not want. IIUC
~~~coq
Definition nat' := nat.
Definition foo T (x:T) := x.
Arguments foo | _ _.

Check (foo nat' : nat -> nat).
(* 
   fresh ?T : Type
   check [foo ?T : nat -> nat] -> ?T := nat 
   check [nat' : Type]
   unify [nat'] and [?T] -> success without side effect 
   return [foo nat : nat -> nat] even though we may expect [foo nat']
 *)
~~~
Is there some way we could fix that?

SkySkimmer commented on pretyping/pretyping.ml:
```diff
@@ -662,20 +683,29 @@ let rec pretype ~program_mode ~poly resolve_tc (tycon : type_constraint) (env :
           let tycon = Some c1 in
-          let sigma, hj = pretype tycon env sigma c in
+          let (sigma, hj), bidiargs =
+            if bidi then
```
Is it useful to do the bidi stuff when there's no tycon? Otherwise we may add `&& tycon <> None` to this condition.

gares commented:
Yes it is fixable at the cost of making the code a bit more convoluted.

I agree about the better guard.

@Janno I'm not so sure I get your suggestion. The pipe signals when the type expected by the context of the entire term has to be used.

JasonGross commented:
I think it might be useful to have an option that automatically sets this point to after the final dependent argument.  (Also, it might make sense to someday flip the argument order of eliminators.)

Janno commented:
> @Janno I'm not so sure I get your suggestion. The pipe signals when the type expected by the context of the entire term has to be used.

Ah, I think I misunderstood the PR then. Sorry for the noise.

Janno commented:
Okay, I will add a bit more noise to be sorry for at some later time:

Is there any chance that this machinery can be used to fix https://github.com/coq/coq/issues/5767? I am a bit too tired to figure out how this differs from Ralf's bug that prompted this PR but I had always assumed these two bugs to be equivalent. Sadly they are not, seeing how mine is not fixed by this PR. (I tried with all possible locations for `&`..)

Are these problems unrelated?

maximedenes commented:
@Janno thanks for mentioning it! This PR also fixes #5767. You can try it with:

```coq
Arguments base {A} m a & _.
Arguments tele {A} m {X} & _.
```

Janno commented:
I am extremely baffled by this and I don't understand how annotations for base and tele can fix the issue but I will gladly take this solution! :)

maximedenes commented:
> Yes it is fixable at the cost of making the code a bit more convoluted.

@gares can you describe the fix you have in mind?

gares commented:
We have to document the feature anyway, so let's start here.

The user writes `f _ b` when `f : forall u, W u -> R u` with `Arguments f _ & _`. The system elaborates `f ?1 ?2` (yes, `b` was put aside and replaced by evar ?2) against an expected type, say `R a`. Given that `f ?1 ?2 : R ?1` the elaborator fixes `?1 = a` by unifying `R a = R ?1`. Now it goes back to the arguments after the `&`: when `b` is elaborated, its expected type is now `W a` instead of `W ?1` since `a` was read from the context, rather than inferred from (the type of) `b`. The difference is that sometimes `W` is a type constructor that recurse on its argument in order to generate a type that actually matches the one of `b`. This is the case for in the example of Ralf.

This feature is a generalization of the one presented in https://arxiv.org/abs/1202.4905 section 4.1. Example 4.5 has a little bug, but the final solution is correct. In that example the type constructor is not recursive, but higher order. Eg `ex_intro ?T ?P ?x p : ex nat (fun x => x > 0)`. Given `p : 2 > 0`, if one type checks the application ignoring the expected type he has to solve the unification problem `?P ?x = 2 > 0` that is very hard. If one says `Arguments ex_intro _ _ & _ _` then he has to solve `(fun x => x > 0) ?x = 2 > 0` that, once the LHS is reduced, it is just first order unification.

gares commented:
> @gares can you describe the fix you have in mind?

The elaborator returns a term. Given the current code it returns the term where arguments after `&` are (eventually assigned) evars, no matter than the user wrote (the example of gaetan passes explicitly a term that is also read from the context). When processing the arguments after `&`, you could put them back in place, instead of just ensuring they unify. It is not trivial since the original evar could have been casted using a coercion, etc... so it may not be easy to find. If you recall, this was what the first implementation draft was doing, then we opted for the simpler one currently implemented. This is why I've mixed feelings about that request, that is reasonable, but complicates so much the code...

SkySkimmer commented:
IIUC we expect the common case to have the remaining arguments be non-dependent, then they should be left alone by the unification/coercion (unless we get into some TC/CS issue?) and will be filled by the user given value.

maximedenes commented:
> This is why I've mixed feelings about that request, that is reasonable, but complicates so much the code...
> I also don't know if it is a problem in practice.

IIRC we even concluded we didn't know how to do it.



> IIUC we expect the common case to have the remaining arguments be non-dependent, then they should be left alone by the unification/coercion (unless we get into some TC/CS issue?) and will be filled by the user given value.

Yes, but the application itself could have been changed (like a coercion to `Funclass` inserted), so just knowning *where* to look for the elaborated remaining arguments is not easy.

Coming back to the example, isn't expected after all that bidirectionality annotations give priority to typing info from the expected type? It's not that arguments are discarded, if they don't unify an error will be triggered.

SkySkimmer commented:
>Yes, but the application itself could have been changed (like a coercion to Funclass inserted), so just knowning where to look for the elaborated remaining arguments is not easy.

I mean in this case the extra evars won't be touched so the final unification will define them using the user value without having to look at the application.

gares commented:
Just to try to converge. The blocking issues are:
- libstack
- doc

RalfJung commented on CHANGES.md:
```diff
@@ -9,6 +9,11 @@ Unreleased changes
+- New annotation in `Arguments` for bidirectionality hints: it is now possible
+  to tell type inference to use type information from the context once the `n`
+  first arguments of an application are known. The syntax is:
+  `Arguments foo x y | z`.
```
Looks like it actually is `&`? (The initial comment in the PR is also wrong.)

Janno commented:
@gares thank you very much for this explanation and the reference to your paper!

gares commented:
@maximedenes can we get an updated on this? I know there are several PR that clean up code touched by this PR, but while this one seems kind of ready the cleanups are not. Can the order of the PRs be switched?

maximedenes commented:
> can we get an updated on this? I know there are several PR that clean up code touched by this PR, but while this one seems kind of ready the cleanups are not. Can the order of the PRs be switched?

As I told you yesterday offline, I started documenting it. I fixed other problems (missing libobject etc), but didn't push yet.

maximedenes commented:
Rebased, libobject and documentation added.

SkySkimmer commented on vernac/vernacentries.ml:
```diff
@@ -1428,10 +1466,22 @@ let vernac_arguments ~section_local reference args more_implicits nargs_for_red
+    end;
+
+  if clear_bidi_hint then
+    Pretyping.clear_bidirectionality_hint sr;
```
What about when we're not `section_local`?

maximedenes replied:
Oh, right, I completely forgot to also port this. Will do, thanks!

maximedenes replied:
Done.

maximedenes commented:
@gares This seems ready to go.
